### ISSUE TITLE
Support for Jakarta EE JSON providers (JSON-P & JSON-B)

### ISSUE THREAD

**Issue by greek1979** [json-path/JsonPath#726]
Good day everyone,

Recently, I started looking around for a Java implementation of Stefan Goessner's JSONPath whitepaper, and stumbled across the "JsonPath" project. Upon reviewing the code, I discovered that Jakarta EE (JSON-P and JSON-B) API is not supported yet, but this is what is needed and used in my enterprise application project; after some consideration, I forked the JsonPath and began the implementation, starting with JSON-P (JSR-342) JSON processing, followed by JSON-B (JSR-367) to provide dynamic databinding between JSON objects and POJO.

There are number of important caveats about the Jakarta EE JSON API specifications:

1. Jakarta EE JSON specs come in "Jakarta EE 8" and "Jakarta EE 9" flavour; both expect Java 8 as bare minimum
2. Both JSON-P and JSON-B are API specs; folks out there are welcome to implement them, and some already did
3. JSON-B depends internally on JSON-P (this is what known implementation do at least)

Here is my understanding of how the above fits into the JsonPath context. With regards to (1) and to make the solution future-proof, it is better to go for Jakarta EE 9 specification; it still requires Java 8, but the API is essentially the extension of Jakarta EE 8 hence no features are lost. The item (2) is a bit more tricky, as Jakarta EE approach is radically different from Gson / JsonOrg / Jackson; the final working application should include both the respective JSON API libs, AND the implementations, one for JSON-P, and one for JSON-B. It is a straightforward task to add a hard dependency on JSON-P/JSON-B API artefacts to JsonPath build scripts; for implementations, each developer would have to make a pick autonomously, and the JakartaEE-JsonPath "glue" code should play along nicely with all.

Notable implementations available as of today are Jakarta Glassfish (JSON-P), Eclipse Yasson (JSON-B), and Apache Johnzon (JSON-P and JSON-B).

Finally, the issue (3) has to be dealt with somehow by the aforementioned "glue" code - the one I intend to work on this week.

Ref: https://javaee.github.io/jsonp/ , http://json-b.net

**Comment by greek1979** [comment]
Responding to myself; IF there is a particular reason why JSON-P or JSON-B should NOT be supported by JsonPath, I would be interested to hear it here.

**Comment by kallestenflo** [comment]
There is no particular reason. No one has implemented a JsonProvider and submitted a PR.

**Comment by greek1979** [comment]
Ok, I am working on this "integration". JSON-P is ready, JSON-B had some issues with existing unit tests. Once I am done, i will raise a PR here. Should not take more than couple of days.

**Comment by greek1979** [comment]
The PR is here.

However, upon running JsonPath unit tests against different JSON-B implementations (remember, with JSON-B aka JSR-367, API and implementation are provided separately), I stumbled upon a well-known Oracle vs Jakarta Foundation package naming issue. Back in 2018, JSON API used the umbrella package ``javax.json``; after all the Java EE code was handed over to The Eclipse Foundation and became known as "Jakarta EE", a decision was made to make Big Bang happen, and to rename all ``javax.*`` packages to ``jakarta.*``, including the ``jakarta.json`` umbrella.

As such, third-party libraries - notably, Apache Johnzon from Apache Software Foundation - are no longer compatible with Jakarta EE 9 API declaration, and will not function until updated accordingly by their maintainers. This may take unpredictable amount of time, however, and I would not rule out a possibility that maintainers of some JSON-B libs may chose NOT to go with the flow and stick to the old Jakarta EE 8 spec (officially frozen since Sep 2019).

TL;DR; If backward compatibility with Jakarta EE 8 specifications will ever be required in this project, JSON-P provider and JSON-B mapper classes from the PR above can be cloned, and Java package references changed to ``javax.json``.

**Comment by greek1979** [comment]
Please close this issue as new functionality described here has been implemented, tested, and merged in #734

**Comment by greek1979** [comment]
Let's close this issue as new functionality described here has been implemented, tested, and merged in #734
